### ISSUE TITLE
Payara (and Glassfish) class loading mechanism does not delegate to boot class loader - causing java.lang.NoClassDefFoundError.

### ISSUE THREAD

**Issue by wheredevel** [payara/Payara#4881]
# Description #
----------

Payara (and Glassfish) class loading mechanism does not delegate to boot class loader.  
Hence, classes provided by `-javaagent:XXX.jar` are not available to the lower levels of the class loading.  

Classes instrumented by the agent may refer to classes residing solely in the XXX.jar - hence causing `java.lang.NoClassDefFoundError`.

For example, an agent, that instruments Oracle JDBC driver residing in the `domain/lib` or `glassfish/lib`, will not be loaded with:
```
java.lang.NoClassDefFoundError: Could not initialize class oracle.jdbc.OracleDriver
```
because the required instrumentation causes the class to refer to some other class in the XXX.jar.

## Expected Outcome

This can be considered a BUG, since other Webcontainers do not suffer from the same problem.  

The expected behavior:  
The affected class loader (from the debug we noticed it's `CurrentBeforeParentClassloader`) should delegate to boot class loader.

## Current Outcome

```
java.lang.NoClassDefFoundError: Could not initialize class oracle.jdbc.OracleDriver
```
because the required instrumentation causes the class to refer to some other class in the XXX.jar.

Clearly, in such case, when agent is bootstrapped, a web application requiring DB connectivity - can no longer connect.

## Steps to reproduce (Only for bug reports) 

0 -** Prepare an agent that instruments, for instance, Oracle JDBC driver with a reference (i.e. field) to a class residing in the agent's jar. Place ojdbc-YYY.jar driver in `domain/lib` or `glassfish/lib`.

1 -**  Start the domain

    ./asadmin start-domain payaradomain

2 -**  Configure an HTTP network listener in the admin console:

[Attach screenshots of Payara's Server console, to illustrate]

3 -**  Make changes to the domain.xml configuration:

	<java-config classpath-suffix="" debug-options="-agentlib:jdwp=transport=dt_socket,server=y,suspend=n,address=9009" system-classpath="">
        <jvm-options>-server</jvm-options>
		<jvm-options>-javaagent:XXX.jar</jvm-options>
	</java-config>

## Environment ##

- **Payara Version**: 3.0+
- **Edition**: ANY
- **JDK Version**: ANY
- **Operating System**: ANY
- **Database**: ANY

**Comment by khasunuma** [comment]
I'd like you to confirm that your consideration is based on following documents: 
https://eclipse-ee4j.github.io/glassfish/docs/latest/application-development-guide/class-loaders.html

**Comment by wheredevel** [comment]
> I'd like you to confirm that your consideration is based on following documents:
> https://eclipse-ee4j.github.io/glassfish/docs/latest/application-development-guide/class-loaders.html

I'm familiar with this document.  
It specifies the Bootstrap class loader at the top of the delegation hierarchy.  
But we don't see the delegation occurring during a debug.  
(The document is general, lacks some details. For example, OSGi class loader, whitelisting classes, etc.)

The document does not refer to the `javaagent` case.  
NewRelic and DataDog notifiers ([recently removed](https://github.com/payara/Payara/releases/tag/payara-server-5.2020.4)) do act though `javaagent` mechanism.  
Either 1) they do not create the "reference to the bootstrapped classes situation", OR  
2) they both bootstrap and, also, place the agent lower on the hierarchy like `asdmin add-library` would do, OR
3) Payara maintains NewRelic and DataDog libraries distributed at the module level.   
We don't have details on that - to allow us confirming the issue 100%.

To summarize, our issue report is based on confronting a real-life failure with Payara/Glassfish.  
We invested a significant time into studying documentation and debugging the issue.

**Comment by pdudits** [comment]
Payara uses Apache Felix as OSGi container. OSGi container requires that all classes that are coming from boot classpath are explicitly specified. The place for this information is `glassfish/config/osgi.properties`.

Enumerate all packages provided by an agent in property `org.osgi.framework.bootdelegation` there.

**Comment by wheredevel** [comment]
> Payara uses Apache Felix as OSGi container. OSGi container requires that all classes that are coming from boot classpath are explicitly specified. The place for this information is `glassfish/config/osgi.properties`.
> 
> Enumerate all packages provided by an agent in property `org.osgi.framework.bootdelegation` there.

@pdudits ,

We tried that.  
We added our packages/classes specifications in `osgi.properties`:
- `org.osgi.framework.bootdelegation`
- `extra-system-packages`
(We also tried modifying `org.osgi.framework.bundle.parent` from `framework` to `boot` (which is default) amidst the comment saying GlassFish can't work this way. Neither worked.)

All that doesn't solve the issue.  

From the debug we see that the resolving of the classes residing in our boot-loaded XXX.jar passes through the OSGi class loader to a higher-level delegate of a delegate ... - CurrentBeforeParentClassLoader is the last in the chain that fails.   

Of course, it could be a faulty OSGi classloader implementation, not looking at the `org.osgi.framework.bootdelegation`, like other Webcontainers do. (We support all the available Webcontainers. For some, indeed, adding booted packages/classes to `org.osgi.framework.bootdelegation` is a requirement. For instance, WebSphere.)

BTW. We're speaking about running GlassFish/Payara on Java 8.

**Comment by pdudits** [comment]
@wheredevel in such case we'll need a reproducer, as this method worked for any other agent libraries I'm aware of (i. e. byteman, New Relic or AppDynamics).

You can also disable the behavior of `CurrentBeforeParentClassLoader` by setting system property `fish.payara.classloading.delegate` to `false`. But since you have trouble with Glassfish as well, this probably won't help as this is payara-specific feature.

**Comment by wheredevel** [comment]
@pdudits ,

We can have a live session. I can demo this failure. Including the debug.

**Comment by OndroMih** [comment]
Hi @wheredevel, requesting a live session is part of Payara Enterprise commercial support. Payara engineers don't provide services like this to the community. Our community support requires that the community users provide as much details as possible in an open way, here in github, including a reproducer that is fairly easy to use and clearly reproducers the problem. Without a reliable and easy to use reproducer, we won't be able to dedicate enough time to investigate and fix this issue and we'll need to close eventually.

If you'd like to get dedicated support for your issues, including remote sessions and private communication, please contact Payara Sales at sales@payara.fish. As a Payara customer, you'd get access to Payara Enterprise with enhanced stability guarantees and production-oriented features as well as access to our support services.

**Comment by wheredevel** [comment]
@OndroMih,  

I'm not requesting, but suggesting a live session.  
I'll see what we can do to provide a "reproducer", as you name it.

**Comment by MeroRai** [comment]
Hi @wheredevel,

Due to the lack of activity form you, I am closing this issue immediately. If the issue is still pressing to you, feel free to verify if it’s applicable in the current release of Payara Server Community edition and proceed to raise a new issue with details of the test reproducer. Many thanks for your understanding.

**Comment by goldfita** [comment]
I may be running into this problem or a closely related one. I tried adding packages to `org.osgi.framework.bootdelegation`, `extra-system-packages`, and `org.osgi.framework.system.packages.extra`. In fact, I can comment out bootdelegation, and it doesn't seem to make any difference. I tried dumping my jar in various directories, hoping the class loader would pick it up.

I inspected the class loader chain for both a .war and an OSGi bundle. For a .war, it looks like the following.

```
WebappClassLoader
DelegatingClassLoader
CurrentBeforeParentClassLoader
APIClassLoader
ExtClassLoader
```
My jar gets picked up by either `CurrentBeforeParentClassLoader` or `APIClassLoader`. For a bundle, the class loaders looks like this:

```
BundleClassLoader
URLClassLoader
ExtClassLoader
```
I don't even see `CurrentBeforeParentClassLoader`. The paths member of URLClassLoader include the following, along with a bunch of paths containing osgi-cache.

```
/var/payara5/glassfish/modules/simple-glassfish-api.jar
/var/payara5/glassfish/osgi/felix/bin/felix.jar
/usr/lib/jvm/java-1.8.0-openjdk-1.8.0.372.b07-2.el9.x86_64/lib/tools.jar
```
There doesn't seem to be a way to set the CLASSPATH from `osgi.properties`, only package names. So either, there should be another class loader in the chain above that reads from one of the special paths, or there should be some way to set the CLASSPATH that passes through to `URLClassLoader`. I tried setting it in `domain.xml`, using java `-cp` flag, using asadmin `--libraries` flag, and the environment variable. None had any effect. Maybe there is a hook somewhere where I can set it at runtime before bundles start to load?

**Comment by goldfita** [comment]
By trial and error, I got the following to work. First set the extension classpath.

`<jvm-options>-Djava.ext.dirs=/usr/lib/jvm/java-1.8.0-openjdk-XXX/jre/lib/ext/:/path/to/<domain>/lib/</jvm-options>`

Then create an empty extension bundle as explained in [this post](https://spring.io/blog/2009/01/19/exposing-the-boot-classpath-in-osgi)  (solution A'). Note that this doesn't seem to work by declaring packages in `osgi.properties`. I discovered when I did it that way that I could see and create the classes in the remote debugger, but the actual code would still throw `NoClassDefFoundError` even though the classes were loaded by the extension classloader.

Unfortunately, this creates a new problem, which I haven't solved. While the classes now load from OSGi bundles, they don't seem to work when annotated from non-OSGi, regular web application. For example, `@BeanParam` no longer works. From inspection, both the classloader for the current class and the thread context have the extension classloader as the root parent, and all the classes in my jar are loaded. The classes are no longer loaded from `CurrentBeforeParentClassLoader`, however this shouldn't be a problem since the root classloader should always be queried at some point no matter which classloader you start from. But for some reason, annotations don't work.

**Comment by goldfita** [comment]
After more experimentation, I figured out that if you create an extension bundle on the system bundle as described in the blog post AND you add your packages to `org.osgi.framework.bootdelegation` AND you change `org.osgi.framework.bundle.parent` to `app` AND you add glassfish to the classpath along with any other jars you need (`<jvm-options>-Djava.class.path=/var/payara5/glassfish/modules/glassfish.jar:/path/to/your/library</jvm-options>`), then boot delegation seems to work. However, I still had the same problem with `@BeanParam` not working correctly. And then you also have to figure out which jars under `modules` you need.
